trl/trainer/dpo_trainer.py

```diff
@@ -290,7 +294,219 @@ def _prepare_deepspeed(self, model: PreTrainedModelWrapper):
+
+        # tokenize the dataset and compute reference logps for evaluation datasets
+        eval_dataset = eval_dataset.map(self.tokenize_batch_element)
+        if self.precompute_ref_log_probs:
+            eval_dataset = eval_dataset.map(
+                self.compute_reference_logps, batch_size=self.args.per_device_eval_batch_size, batched=True
+            )
+        return super().get_eval_dataloader(eval_dataset=eval_dataset)
+
+    def tokenize_batch_element(self, feature, model: Union[PreTrainedModel, nn.Module] = None) -> Dict:
+        """Tokenize a single batch element from a DPO specific dataset.
+
+        At this stage, we don't convert to PyTorch tensors yet; we just handle the truncation
+        in case the prompt + chosen or prompt + rejected responses is/are too long. First
+            we truncate the prompt; if we're still too long, we truncate the chosen/rejected.
+
+        We also create the labels for the chosen/rejected responses, which are of length equal to
+            the sum of the length of the prompt and the chosen/rejected response, with
+            label_pad_token_id  for the prompt tokens.
+        """
+        batch = {}
+        prompt = feature["prompt"]
+        chosen = feature["chosen"]
+        rejected = feature["rejected"]
+
+        if not self.is_encoder_decoder:
+            chosen_tokens = self.tokenizer(chosen, add_special_tokens=False)
+            rejected_tokens = self.tokenizer(rejected, add_special_tokens=False)
+            prompt_tokens = self.tokenizer(prompt, add_special_tokens=False)
+
+            eos_token_id = self.tokenizer.eos_token_id
+            # Get indices in list prompt_tokens["input_ids"] that equals the EOS token (often 0)
+            eos_indices_prompt = [i for i, x in enumerate(prompt_tokens["input_ids"]) if x == eos_token_id]
+            # attention mask these indices to eos_token_id
+            new_attention_mask = [
+                0 if i in eos_indices_prompt else p for i, p in enumerate(prompt_tokens["attention_mask"])
+            ]
+            prompt_tokens["attention_mask"] = new_attention_mask
+
+            # do the same for chosen and rejected
+            eos_indices_chosen = [i for i, x in enumerate(chosen_tokens["input_ids"]) if x == eos_token_id]
+            new_attention_mask_c = [
+                0 if i in eos_indices_chosen else p for i, p in enumerate(chosen_tokens["attention_mask"])
+            ]
+            chosen_tokens["attention_mask"] = new_attention_mask_c
+
+            eos_indices_rejected = [i for i, x in enumerate(rejected_tokens["input_ids"]) if x == eos_token_id]
+            new_attention_mask_r = [
+                0 if i in eos_indices_rejected else p for i, p in enumerate(rejected_tokens["attention_mask"])
+            ]
+            rejected_tokens["attention_mask"] = new_attention_mask_r
+
+            # add EOS token to end of prompt
+            chosen_tokens["input_ids"].append(self.tokenizer.eos_token_id)
```

Comment:
For my understanding, why do we add the EOS token at the end of the prompt? For context, when we use chat templates one often has a structure like this:

```
<|system|>
{{ system_prompt }}</s>
<|user|>
{{ instruction }}</s>
<|assistant|>
{{ model_answer }}</s>
```

where the EOS token `</s>` indicates the end of a turn. My understanding of the DPO implementation is that you have a triple like `(prompt, chosen_completion, rejected_completion)`, so my question is whether users should be removing the EOS token from the user turn of their chat templates or not

Reply:
> For my understanding, why do we add the EOS token at the end of the prompt? For context, when we use chat templates one often has a structure like this:
> 
> ```
> <|system|>
> {{ system_prompt }}</s>
> <|user|>
> {{ instruction }}</s>
> <|assistant|>
> {{ model_answer }}</s>
> ```
> 
> where the EOS token `</s>` indicates the end of a turn. My understanding of the DPO implementation is that you have a triple like `(prompt, chosen_completion, rejected_completion)`, so my question is whether users should be removing the EOS token from the user turn of their chat templates or not

I think the  chosen_tokens here is the concatenated inputs of 'prompt + chosen', instead of the prompt only. please check if it is right or not? I think the EOS token is append to the end of the whole "prompt + chosen_completion" here